src/challengePaths/2021/YouRobot.ts

```diff
@@ -0,0 +1,139 @@
+import { round, runChoice, Slot, Stat, visitUrl } from "kolmafia";
+import { have } from "../../lib.js";
+import { get } from "../../property.js";
+import { $path, $skill, $slot, $stat } from "../../template-string.js";
+
+export const path = $path`You, Robot`;
+
+type Upgrade = "robot_shirt" | "robot_energy" | "robot_potions";
+
+/**
+ * Runs one turn scavenging in the scrapheap
+ */
+export function scavenge(): void {
+  visitUrl("place.php?whichplace=scrapheap&action=sh_scrounge");
+}
+
+/**
+ * Spends one turn collecting energy in the scrapheap
+ */
+export function collectEnergy(): void {
+  visitUrl("place.php?whichplace=scrapheap&action=sh_getpower");
+}
+
+/**
+ * @returns the energy cost for the next use of the chronolith
+ */
+export function expectedEnergyNextCollect(): number {
+  const raw = (25 + get("youRobotPoints")) * 0.85 ** get("_energyCollected");
+  return round(raw);
+}
+
+/**
+ * Spends energy at the chronolith to get 10 adventures
+ */
+export function visitChronolith(): void {
+  visitUrl("place.php?whichplace=scrapheap&action=sh_chronobo");
+}
+
+/**
+ * @returns how much energy we expect the next chronolith use to take
+ */
+export function expectedChronolithCost(): number {
+  return get("_chronolithNextCost");
+}
+
+/**
+ *
+ * @param stat the stat we want to boost using the statbot
+ */
+export function visitStatbot(stat: Stat) {
+  visitUrl("place.php?whichplace=scrapheap&action=sh_upgrade");
+  if (stat === $stat`Muscle`) runChoice(1);
+  else if (stat === $stat`Mysticality`) runChoice(2);
+  else if (stat === $stat`Moxie`) runChoice(3);
+}
+/**
+ * @param uses defines the number of times we want to visit the statbot
+ * @returns the anticipated cost of the total visits
+ */
+export function expectedStatbotCost(uses = 1): number {
+  return (10 + get("statbotUses")) * uses + ((uses - 1) * uses) / 2;
+}
+
+/**
+ * @param upgrade defines the robot part you want to check for
+ * @returns whether that part is currently equipped
+ */
+export function haveUpgrade(upgrade: Upgrade): boolean {
+  return get("youRobotCPUUpgrades").includes(upgrade);
+}
+
+/**
+ * @param upgrade defines the upgrade you want to complete
+ * @returns whether the upgrade was successfully completed
+ */
+export function doUpgrade(upgrade: Upgrade): boolean {
+  if (haveUpgrade(upgrade)) return true;
+  visitUrl("place.php?whichplace=scrapheap&action=sh_configure");
+  visitUrl("choice.php?whichchoice=1445&show=cpus");
+  visitUrl(
+    `choice.php?pwd&whichchoice=1445&part=cpus&show=cpus&option=2&p=${upgrade}`,
+  );
+  return haveUpgrade(upgrade);
+}
+
+function getPartId(which: "top" | "left" | "right" | "bottom") {
+  switch (which) {
+    case "top":
+      return get("youRobotTop");
+    case "left":
+      return get("youRobotLeft");
+    case "right":
+      return get("youRobotRight");
+    case "bottom":
+      return get("youRobotBottom");
+  }
+}
```

Comment:
```suggestion
function getPartId(slot: "Top" | "Left" | "Right" | "Bottom") {
  return get(`youRobot${part}`);
}
```

Reply:
That being said, this name to me suggests it's a function where you input a _part_, and get the id number associated with that.

I'd consider providing Enums or something like that to allow people to refer to specific parts by name, instead of having to know their ids off the top of their head.